msysmon.8
.\"
.\" Man page for Mike's System Monitor.
.\"
.\" Copyright © 2026 by Michael R Sweet
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH msysmon 1 "msysmon" "2026-08-10" "Michael R Sweet"
.SH NAME
msysmon \- mike's system monitor
.SH SYNOPSIS
.B msysmon
[
.B \-\-help
] [
.B \-\-version
] [
.B \-C
.I CPU-LIMIT
] [
.B \-I
.I INTERVAL
] [
.B \-M
.I MEMORY-LIMIT
] [
.B \-p
.I PORT-NUMBER
] [
.B \-s
.I FIELD
] [
.B \-w
.I COMMAND
]
.SH DESCRIPTION
.B msysmon
monitors the running system and its processes to identify trends and peak users
of CPU, memory, network, and storage on the system.
By default,
.B msysmon
provides a simple web interface to view up to the last four weeks of system
usage and download the data as CSV files for offline processing.
.SH OPTIONS
.TP 5
.B \-\-help
Shows program usage.
.TP 5
.B \-\-version
Shows program version.
.TP 5
\fB\-C \fICPU-LIMIT\fR
Specifies the minimum CPU usage (as a percentage) that will trigger monitoring of a process.
The default is 25%.
.TP 5
\fB\-I \fIINTERVAL\fR
Specifies the sampling interval.
The default is 300 seconds (5 minutes).
The interval can use the unit suffix "s" for seconds (the default unit), "m" for minutes, "h" for hours, or "d" for days.
.TP 5
\fB\-M \fIMEMORY-LIMIT\fR
Specifies the minimum virtual memory usage that will trigger monitoring of a process.
The default is 5% of available real system memory.
The limit can use the unit suffix "%" for a percentage of real memory, "k" for multiples of 1,024 bytes, "m" for multiples of 1,048,576 bytes, or "g" for multiples of 1,073,741,824 bytes.
.TP 5
\fB\-p \fIPORT-NUMBER\fR
Specifies the port number of the web interface.
The default post number is 8080.
.TP 5
\fB\-s \fIFIELD\fR
Specifies the primary sorting field for the process list in the web interface - "cpu", "mem", "name", "pid", or "threads".
A leading "-" specifies descending order.
The default is "-cpu".
.TP 5
\fB\-w \fICOMMAND\fR
Explicitly watches the specified command.
By default
.B msysmon
watches any process that exceeds the specified CPU or memory limits.
.SH SEE ALSO
.BR proc (5)
.SH COPYRIGHT
Copyright \[co] 2026 by Michael R Sweet.


CHANGES.md
Changes in Mike's System Monitor
================================

v1.0.0 - YYYY-MM-DD
-------------------

- Initial public release.


msysmon.c
//
// Main entry for Mike's System Monitor.
//
// Copyright © 2026 by Michael R Sweet
//
// Licensed under Apache License v2.0.  See the file "LICENSE" for more
// information.
//

#include "msysmon.h"
#include <cups/dnssd.h>


//
// Globals...
//

msysmon_db_t	msysmonData;		// Monitoring data/settings


//
// Local functions...
//

static void	dnssd_cb(cups_dnssd_service_t *service, void *cb_data, cups_dnssd_flags_t flags);
static int	usage(FILE *out);


//
// 'main()' - Main entry.
//

int					// O - Exit status
main(int  argc,				// I - Number of command-line arguments
     char *argv[])			// I - Command-line arguments
{
  int			i;		// Looping var
  const char		*opt;		// Current option
  double		val;		// Value
  char			*end;		// Pointer into value
  char			host[256],	// Hostname
			port[255],	// Service port
			*nameptr;	// Pointer into service name
  http_addrlist_t	*addrlist;	// Listen address
  time_t		collect_time;	// Next data collection time
  cups_dnssd_t		*dnssd;		// DNS-SD context
  cups_dnssd_service_t	*service;	// DNS-SD service


  // Initialize data...
  memset(&msysmonData, 0, sizeof(msysmonData));

  cupsRWInit(&msysmonData.rwlock);

  msysmonData.interval  = DEFAULT_INTERVAL;
  msysmonData.port      = DEFAULT_PORT;
  msysmonData.cpu_limit = DEFAULT_CPU_LIMIT;
  msysmonData.mem_limit = DEFAULT_MEM_LIMIT * msysmonGetSystemMemory() / 100;
  msysmonData.sort_by   = -FIELD_CPU;

  // Parse command-line...
  for (i = 1; i < argc; i ++)
  {
    if (!strcmp(argv[i], "--help"))
    {
      return (usage(stdout));
    }
    else if (!strcmp(argv[i], "--version"))
    {
      puts(VERSION);
      return (0);
    }
    else if (!strncmp(argv[i], "--", 2) || argv[i][0] != '-')
    {
      fprintf(stderr, "msysmon: Unknown option '%s'.\n", argv[i]);
      return (usage(stderr));
    }
    else
    {
      for (opt = argv[i] + 1; *opt; opt ++)
      {
        switch (*opt)
        {
          case 'C' : // -C CPU-LIMIT
              i ++;
              if (i >= argc)
              {
                fputs("msysmon: Missing CPU limit after '-C'.\n", stderr);
                return (usage(stderr));
              }

	      if ((val = strtod(argv[i], &end)) < 1 || val > 100 || *end)
	      {
	        fprintf(stderr, "msysmon: Bad CPU limit '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }

	      msysmonData.cpu_limit = (int)val;
              break;

          case 'I' : // -I INTERVAL
              i ++;
              if (i >= argc)
              {
                fputs("msysmon: Missing sampling interval after '-I'.\n", stderr);
                return (usage(stderr));
              }

	      if ((val = strtod(argv[i], &end)) < 0)
	      {
	        fprintf(stderr, "msysmon: Bad sampling interval '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }
	      else if (*end == 'm')
	      {
	        msysmonData.interval = (int)(val * 60.0);
	      }
	      else if (*end == 'h')
	      {
	        msysmonData.interval = (int)(val * 60.0 * 60.0);
	      }
	      else if (*end == 'd')
	      {
	        msysmonData.interval = (int)(val * 24.0 * 60.0 * 60.0);
	      }
	      else if (*end && *end != 's')
	      {
	        fprintf(stderr, "msysmon: Bad sampling interval '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }
	      else
	      {
	        msysmonData.interval = (int)val;
	      }

	      if (msysmonData.interval == 0)
	      {
	        fputs("msysmon: Sampling interval cannot be less than one second.\n", stderr);
	        return (usage(stderr));
	      }
              break;

          case 'M' : // -M MEM-LIMIT
              i ++;
              if (i >= argc)
              {
                fputs("msysmon: Missing memory limit after '-M'.\n", stderr);
                return (usage(stderr));
              }

	      if ((val = strtod(argv[i], &end)) < 1)
	      {
	        fprintf(stderr, "msysmon: Bad memory limit '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }
	      else if (*end == '%')
	      {
	        msysmonData.mem_limit = (uint32_t)(val * msysmonGetSystemMemory() / 100.0);
	      }
	      else if (*end == 'k')
	      {
	        msysmonData.mem_limit = (uint32_t)val;
	      }
	      else if (*end == 'm')
	      {
	        // Convert mebibytes to kibibytes
	        msysmonData.mem_limit = (uint32_t)(val * 1024.0);
	      }
	      else if (*end == 'g')
	      {
	        // Convert gibibytes to kibibytes
	        msysmonData.mem_limit = (uint32_t)(val * 1024.0 * 1024.0);
	      }
	      else if (*end)
	      {
	        fprintf(stderr, "msysmon: Bad memory limit '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }
	      else
	      {
	        // Convert bytes to kibibytes
	        msysmonData.mem_limit = (uint32_t)(val / 1024.0);
	      }
              break;

          case 'p' : // -p PORT-NUMBER
              i ++;
              if (i >= argc)
              {
                fputs("msysmon: Missing port number after '-p'.\n", stderr);
                return (usage(stderr));
              }

	      if ((val = strtod(argv[i], &end)) < 1 || val > 65535 || *end)
	      {
	        fprintf(stderr, "msysmon: Bad port number '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }

	      msysmonData.port = (int)val;
              break;

	  case 's' : // -s FIELD
              i ++;
              if (i >= argc)
              {
                fputs("msysmon: Missing command after '-w'.\n", stderr);
                return (usage(stderr));
              }

	      if (!strcasecmp(argv[i], "cpu") || !strcasecmp(argv[i], "+cpu"))
	        msysmonData.sort_by = FIELD_CPU;
	      else if (!strcasecmp(argv[i], "-cpu"))
	        msysmonData.sort_by = -FIELD_CPU;
	      else if (!strcasecmp(argv[i], "mem") || !strcasecmp(argv[i], "+mem"))
	        msysmonData.sort_by = FIELD_MEM;
	      else if (!strcasecmp(argv[i], "-mem"))
	        msysmonData.sort_by = -FIELD_MEM;
	      else if (!strcasecmp(argv[i], "name") || !strcasecmp(argv[i], "+name"))
	        msysmonData.sort_by = FIELD_NAME;
	      else if (!strcasecmp(argv[i], "-name"))
	        msysmonData.sort_by = -FIELD_NAME;
	      else if (!strcasecmp(argv[i], "pid") || !strcasecmp(argv[i], "+pid"))
	        msysmonData.sort_by = FIELD_PID;
	      else if (!strcasecmp(argv[i], "-pid"))
	        msysmonData.sort_by = -FIELD_PID;
	      else if (!strcasecmp(argv[i], "threads") || !strcasecmp(argv[i], "+threads"))
	        msysmonData.sort_by = FIELD_THREADS;
	      else if (!strcasecmp(argv[i], "-threads"))
	        msysmonData.sort_by = -FIELD_THREADS;
	      else
	      {
	        fprintf(stderr, "msysmon: Unknown sort by field '%s'.\n", argv[i]);
	        return (usage(stderr));
	      }
              break;

          case 'w' : // -w COMMAND
              i ++;
              if (i >= argc)
              {
                fputs("msysmon: Missing command after '-w'.\n", stderr);
                return (usage(stderr));
              }

	      if (msysmonData.num_commands < MAX_COMMANDS)
	      {
	        cupsCopyString(msysmonData.commands[msysmonData.num_commands], argv[i], sizeof(msysmonData.commands[0]));
	        msysmonData.num_commands ++;
	      }
	      else
	      {
	        fputs("msysmon: Too many commands to watch.\n", stderr);
	        return (usage(stderr));
	      }
              break;

          default :
              fprintf(stderr, "msysmon: Unknown option '-%c'.\n", *opt);
              return (usage(stderr));
        }
      }
    }
  }

  MSYSMON_DEBUG("sizeof(msysmon_db_t) = %lu\n", (unsigned long)sizeof(msysmonData));
  MSYSMON_DEBUG("interval = %d\n", msysmonData.interval);
  MSYSMON_DEBUG("cpu_limit = %u\n", msysmonData.cpu_limit);
  MSYSMON_DEBUG("mem_limit = %u\n", msysmonData.mem_limit);

  // Create listeners...
  httpGetHostname(/*http*/NULL, host, sizeof(host));
  snprintf(port, sizeof(port), "%d", msysmonData.port);

  if ((addrlist = httpAddrGetList(/*name*/NULL, AF_INET, port)) == NULL)
  {
    fprintf(stderr, "msysmon: Unable to lookup IPv4 listen address: %s\n", cupsGetErrorString());
    return (1);
  }
  else
  {
    if ((msysmonData.listeners[msysmonData.num_listeners].fd = httpAddrListen(&(addrlist->addr), msysmonData.port)) >= 0)
    {
      msysmonData.listeners[msysmonData.num_listeners].events = POLLIN | POLLERR | POLLHUP;
      msysmonData.num_listeners ++;
    }

    httpAddrFreeList(addrlist);
  }

  if ((addrlist = httpAddrGetList(/*name*/NULL, AF_INET6, port)) == NULL)
  {
    fprintf(stderr, "msysmon: Unable to lookup IPv4 listen address: %s\n", cupsGetErrorString());
    return (1);
  }
  else
  {
    if ((msysmonData.listeners[msysmonData.num_listeners].fd = httpAddrListen(&(addrlist->addr), msysmonData.port)) >= 0)
    {
      msysmonData.listeners[msysmonData.num_listeners].events = POLLIN | POLLERR | POLLHUP;
      msysmonData.num_listeners ++;
    }

    httpAddrFreeList(addrlist);
  }

  if (msysmonData.num_listeners == 0)
  {
    fputs("msysmon: Unable to create IPv4/6 listeners.\n", stderr);
    return (1);
  }

  // Advertise the web page...
  fprintf(stderr, "msysmon: Web interface available at 'http://%s:%d/'.\n", host, msysmonData.port);

  cupsCopyString(msysmonData.name, host, sizeof(msysmonData.name));
  if ((nameptr = strchr(msysmonData.name, '.')) != NULL)
    *nameptr = '\0';
  cupsConcatString(msysmonData.name, " Statistics", sizeof(msysmonData.name));

  if ((dnssd = cupsDNSSDNew(/*error_cb*/NULL, /*cb_data*/NULL)) != NULL)
  {
    bool	success = false;		// Did we advertise successfully?

    if ((service = cupsDNSSDServiceNew(dnssd, CUPS_DNSSD_IF_INDEX_ANY, msysmonData.name, dnssd_cb, /*cb_data*/NULL)) != NULL)
    {
      success = cupsDNSSDServiceAdd(service, "_http._tcp", /*domain*/NULL, /*host*/NULL, msysmonData.port, /*num_txt*/0, /*txt*/NULL);
      if (success)
        success = cupsDNSSDServicePublish(service);
    }

    if (!success)
      fprintf(stderr, "msysmon: Unable to advertise web interface as '%s'.\n", msysmonData.name);
  }
  else
  {
    fputs("msysmon: Unable to advertise web interface.\n", stderr);
  }

  // Run...
  msysmonData.data_start = time(NULL);
  msysmonData.data_start -= msysmonData.data_start % msysmonData.interval;
  collect_time           = msysmonData.data_start + msysmonData.interval;

  if (!msysmonCollectData())
    return (1);

  for (;;)
  {
    if (time(NULL) >= collect_time)
    {
      if (!msysmonCollectData())
        break;

      collect_time += msysmonData.interval;
    }

    if (poll(msysmonData.listeners, msysmonData.num_listeners, 1000) > 0)
    {
      // Accept new connections...
      for (i = 0; i < msysmonData.num_listeners; i ++)
      {
        if (msysmonData.listeners[i].revents & POLLIN)
        {
	  MSYSMON_DEBUG("Accepting connection from socket %u...\n", i);
	  http_t *http = httpAcceptConnection(msysmonData.listeners[i].fd, /*blocking*/true);
					// New connection

          if (http)
          {
	    MSYSMON_DEBUG("Starting web interface thread for new connection...\n");
            cups_thread_t tid = cupsThreadCreate((cups_thread_func_t)msysmonRunWebIf, http);
            cupsThreadDetach(tid);
          }
        }
      }
    }
  }

  return (0);
}


//
// 'dnssd_cb()' - DNS-SD service callback.
//

static void
dnssd_cb(
    cups_dnssd_service_t *service,	// I - Service
    void                 *cb_data,	// I - Callback data (unused)
    cups_dnssd_flags_t   flags)		// I - Flags
{
  MSYSMON_DEBUG("DNS-SD callback for service %p (%s), flags=0x%02X\n", service, cupsDNSSDServiceGetName(service), flags);

  (void)service;
  (void)cb_data;
  (void)flags;
}


//
// 'usage()' - Show program usage.
//

static int				// O - Exit status
usage(FILE *out)			// I - Output file
{
  fputs("Usage: msysmon [OPTIONS]\n", out);
  fputs("Options:\n", out);
  fputs("  --help             Show help.\n", out);
  fputs("  --version          Show version.\n", out);
  fputs("  -C CPU-LIMIT       Set CPU percentage for auto-monitoring.\n", out);
  fputs("  -I INTERVAL        Set sampling interval.\n", out);
  fputs("  -M MEM-LIMIT       Set memory usage for auto-monitoring.\n", out);
  fputs("  -p PORT-NUMBER     Set web interface port.\n", out);
  fputs("  -s [-]FIELD        Sort the process list by cpu, mem, name, pid, or threads.\n", out);
  fputs("  -w COMMAND         Add command to monitor.\n", out);

  return (out == stdout ? 0 : 1);
}
